Components/Terrain/src/OgreTerrainGroup.cpp

```diff
@@ -825,6 +837,26 @@ namespace Ogre
 
     }
     //---------------------------------------------------------------------
+    void TerrainGroup::freeTerrainSlotInstance(TerrainSlot* slot)
+    {
+        if (!slot)
+            return;
+
+        TerrainLoadRequestMap::iterator it = mTerrainLoadRequests.find(slot->instance);
+
+        // Terrain was in load request so we need to schedule the deletion see handleResponse().
+        if (it != mTerrainLoadRequests.end())
+        {
+            // We would like to abort the request here but WorkQueue does not provide proper tools for this
+            // (the data we need will be destroyed if we call abortRequest()).
```

Comment:
which data are you referring to here? As far as I can see `abortRequest()` will only delete `ResourceRequest::loadParams`

Reply:
You are probably referring to the `TerrainSlot*`. Would calling abortRequest() and immediately deleting it here work?

Reply:
You're correct that the problem is `TerrainSlot*`. It can't be deleted however since `Terrain*` would also be deleted and that is used in the thread. If `WorkQueue` provided a method to abort request if it was not being processed right now it would be fine. That way we would have tools to abort all other requests except the one that's currently being processed.

Reply:
`WorkQueue` has function `abortPendingRequestsByChannel(uint16 channel)` so if there was function like  `abortPendingRequest(RequestID id)` as well it could be done.

Reply:
> if there was function like abortPendingRequest(RequestID id) as well it could be done.

this will probably be useful for others as well. So I would prefer this solution over the currently proposed workaround.

Reply:
I'll make this change but currently I'm a bit busy so it might take a few days.